util/gpool/spmc/spmcpool.go

```diff
@@ -0,0 +1,395 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spmc
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/util/gpool"
+	"go.uber.org/zap"
+)
+
+// Pool is a single producer, multiple consumer goroutine pool.
+type Pool[T any, U any, C any, CT any, TF gpool.Context[CT]] struct {
+	workerCache   sync.Pool
+	lock          sync.Locker
+	workers       workerArray[T, U, C, CT, TF]
+	taskCh        chan *gpool.TaskBox[T, U, C, CT, TF]
+	options       *Options
+	stopCh        chan struct{}
+	consumerFunc  func(T, C, CT) U
+	cond          *sync.Cond
+	name          string
+	taskManager   gpool.TaskManager[T, U, C, CT, TF]
+	generator     atomic.Uint64
+	capacity      atomic.Int32
+	running       atomic.Int32
+	state         atomic.Int32
+	waiting       atomic.Int32
+	heartbeatDone atomic.Int32
+}
+
+// NewSPMCPool create a single producer, multiple consumer goroutine pool.
+func NewSPMCPool[T any, U any, C any, CT any, TF gpool.Context[CT]](name string, size int32, options ...Option) (*Pool[T, U, C, CT, TF], error) {
```

Comment:
Could we remove the `error` in the return? Then we need to check the error when calling this function.